### ISSUE TITLE
Enable usage message to provide runtime documentation of configuration options

### ISSUE THREAD

**Issue by davidkbainbridge** [kelseyhightower/envconfig#45]
Most applications today provide a `-h` or `--help` command line argument that will display the usage of the command and describe, briefly, the available command line arguments and their default values. 

If an application is configured by its environment there should be a mechanism by which a user can be provided with a usage description at runtime. 

(see discussion for more information: https://github.com/kelseyhightower/envconfig/issues/40)

**Comment by davidkbainbridge** [comment]
Just submitted a pull request that implement this https://github.com/kelseyhightower/envconfig/pull/52.

It doesn't capture the --help flags, but supports the output (io.Writer) for the message so when to output is controlled by the app. Also the format can be customized as a Go template.

lastly, and really the biggest change, is i refactored the code so that it follows a visitor pattern so the help message and the processing from environment to structs values are driven by the same loop, just different functions that are being called on each item.

**Comment by davidkbainbridge** [comment]
@teepark: continuing conversation. did you get a chance to think and elaborate on how you would like to see a solution to this issue implemented. 

I agree the code needs to be approachable, but would like to hear more about your thoughts.

**Comment by teepark** [comment]
So right now I think the set of requirements we're working from looks roughly like:
- avoid duplicating code (fears of drift between the copies)
- general enough to provide a head start on related features (such as displaying current values as well)
- maintain general logical structure (no inversion of control flow, etc)

I had in mind breaking apart 2 things that `Process` currently does:
- pull relevant tags from the `spec interface{}` and other per-field info that will drive behavior
- grab values from the environment and set them in the struct

Then the first part would be reusable elsewhere in the package. Something rougly like:

```go
type varInfo struct {
    prefix     string
    name       string // envconfig tag, fallback to field name
    value      reflect.Value
    defaultVal string
    // ...other stuff from supported struct field tags
}

func gatherInfo(prefix string, spec interface{}) ([]varInfo, error) {
    // this may have to be recursive to get child structs, like how Process is today
}

func Process(prefix string, spec interface{}) error {
    infos, err := gatherInfo(prefix, spec)
    for _, info := range infos {
        // populate info.value from environment var
    }
}

func Usage(tmpl string, prefix string, spec interface{}) (string, error) {
    tplate, err := template.New("").Parse(tmpl)
    if err != nil {
        return err
    }

    infos, err := gatherInfo(prefix, spec)
    if err != nil {
        return err
    }

    buf := new(bytes.Buffer)
    err = tplate.Execute(buf, infos)
    return buf.String(), err
}
```

Does that work for you?

**Comment by davidkbainbridge** [comment]
My concern with this approach, and my previous patch has the same problem I recall, is that this design assumes that the `gather` func knows all possible structure tags that are required. 

what about if the `varInfo` essentially processed that tags into a map[string][string] so that each method, such as `Process` or `Usage` can process only the tags they care about.

Additionally, in the proposed design, does this assume that the values are read from the environment even if they are not required by the rest of the processing loop? This would mean extra and unneeded processing in the case of `Usage` for example.

The basic structure works for me, the question is what is populated in the `varInfo`. This solution also means two passes over each configuration attribute (once for gather and once for other processing).

**Comment by teepark** [comment]
> this design assumes that the gather func knows all possible structure tags that are required
>
> what about if the varInfo essentially processed that tags into a map[string][string]

I think it's pretty reasonable to nail down the static list of envconfig-involved struct tags. A simple thing to do though could be just dropping the `reflect.StructTag` right into the `varInfo`, I don't see anything wrong with that.

> does this assume that the values are read from the environment even if they are not required by the rest of the processing loop?

No, `gatherInfo` shouldn't need to read any environment variables. `Process()` would do that, getting the env var names from the data in the `[]varInfo`.

> This solution also means two passes over each configuration attribute (once for gather and once for other processing).

Yep, it does mean two passes over each field.

**Comment by davidkbainbridge** [comment]
submitted change set #62 that implements the two phase approach. please review at your convenience.

**Comment by davidkbainbridge** [comment]
ping. i submitted a changeset pending your review when you get a chance.
thx.

**Comment by teepark** [comment]
Resolved with #62.
